### PR TITLE
Mirrors now display your reflection.

### PR DESCRIPTION
**Pull request by spessbandit**
## About The Pull Request

>[qustinnus] God seeing this being discussed as a possibility on Ss14 made me cum

mirrors

Mirrors now display the reflection of any living mob in a 2-space radius. Currently accomplished with some vis_contents and matrix fuckery that I am open to suggestions on how to make less janky.

**TODO:**
* Figure out how to make north/south-facing mirrors rotate the sprite properly, currently they do not
* Make sure nothing weird happens with magic mirrors or broken mirrors
* Add vampire check so vamps don't appear in the mirror (and/or morphed morphs?)
* Add a phobia for seeing your own reflection
* vanity/dysmorphia traits/moodlets?
* turn into a reflectivity component with vars/arguments for mask sprite, alpha level and scalars?
* did mirrors ever become craftable/deconnable

## Why It's Good For The Game

Adds realism.

## Changelog
:cl: bandit
add: Nanotrasen has upgraded mirrors to actually function as mirrors.
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
Idea: Make the description of a mirror temporarily become "Despite everything, it's still you." if you examine it as an antag over an hour into the round (or "It's you!" if you examine it as a non-antag over an hour into the round). By "temporarily", I mean it just displays that text instead of the normal description of the mirror for that examination only, so people can't use it as an antag check.

**Comment by Fikou** [comment]
> Idea: Make the description of a mirror temporarily become "Despite everything, it's still you." if you examine it as an antag over an hour into the round (or "It's you!" if you examine it as a non-antag over an hour into the round). By "temporarily", I mean it just displays that text instead of the normal description of the mirror for that examination only, so people can't use it as an antag check.

sand undertal

**Comment by spessbandit** [comment]
@ATH1909 that's already on local, or similar; I just used 30 minutes flat

**Comment by Qustinnus** [comment]
this should be modularized so we can have window reflections like ss14 wants

**Comment by Ratvar** [comment]
...maybe revenants

**Comment by Dennok** [comment]
add cursed mirrors that irror ghosts

File: code/game/objects/structures/mirror.dm [hunk 1]
```diff
@@ -54,12 +61,51 @@
...
+	RegisterSignal(AM, COMSIG_MOVABLE_MOVED, .proc/check_reflection, override = TRUE)
+	reflection.vis_contents += AM
+	reflection.filters += filter(type="alpha", icon = icon(src.icon, "mirror_mask"))
+	var/matrix/M = matrix()
+	M.Scale(0.75)
+	M.Multiply(matrix(-1, 0, 0, 0, 1, 0)) // reflection matrix
+	reflection.transform = M
+	vis_contents += reflection
+	if(AM.GetComponent(/datum/component/mood) && AM.GetComponent(/datum/component/mood).sanity <= SANITY_DISTURBED)
+		desc = "Despite everything, it's still you."
```
Comment (PKPenguin321) [inline code/game/objects/structures/mirror.dm]: but what if someone else is sntaind gcloser to the mirror and its reflectign them so its actually not you

**Comment by Featherfield** [comment]
You know only reason vampires were not seen in mirrors in olden days is because they had LARGE AMOUNT OF SILVER in them.
just because its reflective doesnt mean they wont be seen get your lore right retard.

**Comment by PKPenguin321** [comment]
@Featherfield 
### 🚨 👮 🚓 🚔 ⚠️ 🚨 R-WORD DETECTED, GITHUB COPS ARE EN-ROUTE. EDIT YOUR POST OR BE KILLED IMMEDIATELY BY OUR T-O-S-9000 🚨 ⚠️ 🚔 🚓 👮 🚨

**Comment by spessbandit** [comment]
hi it's me my life has been a clusterfuck in the past few days because of work and school being fucked by the coronavirus so it appears I'll have a lot of time to work on this soon!!

**Comment by spookydonut** [comment]
> 
> 
> hi it's me my life has been a clusterfuck in the past few days because of work and school being fucked by the coronavirus so it appears I'll have a lot of time to work on this soon!!

make north/south mirrors just do the scale transform but don't flip, should be close enough.
the only other way to do it would require getflaticon(), though I could see just caching north/south states for mobs and only updating them in spare cpu time.

File: code/game/objects/structures/mirror.dm [hunk 2]
```diff
@@ -54,12 +61,51 @@
...
 		broken = TRUE
 
+/**
+  * If a (living, non-ghost) mob is within 2 tiles of the mirror, their reflection is displayed.
+  *
+  * The mob is added to the mirror container's vis_contents, which is scaled down, reflected, and alpha masked.
+  * The mirror container in turn has the reflection in its vis_contents.
+  */
+
+/obj/structure/mirror/HasProximity(atom/movable/AM as mob|obj)
```
Comment (Rohesie) [inline code/game/objects/structures/mirror.dm]:
```suggestion
/obj/structure/mirror/HasProximity(atom/movable/AM)

```

File: code/game/objects/structures/mirror.dm (hunk 1 above)
Comment (Rohesie) [reply code/game/objects/structures/mirror.dm]: Also, use a signal instead of `GetComponent`

**Comment by spessbandit** [comment]
update: am reformatting my pc for unrelated reasons, will get back to this once I do

**Comment by Cyberboss** [comment]
Stale

**Pull request closed without merging**